### PR TITLE
Include Encryption In-Transit state as part of the desired state hash sent to clients

### PR DESCRIPTION
**Pull request by malayparida2000**
When in-transit encryption is enabled/disabled the kernel mount option for cephFS needs to be updated between prefer-crc/secure. So the desired state hash needs to include the EiT state, so that if the EiT state is changed the desired state hash will change and the client will reconcile to get the updated mount option.

### CODE REVIEW

**Comment by malayparida2000** [comment]
Should go after https://github.com/red-hat-storage/ocs-operator/pull/2793

File: services/provider/server/server.go
Comment (nb-ohad) [inline services/provider/server/server.go:891]: This pattern repeats itself. I would suggest to replace the `getStorageCluster` utils with a `IsEncryptionInTransitEnabled()` function and call it in both places

File: services/provider/server/server.go
Comment (nb-ohad) [inline services/provider/server/server.go:944]: 1. Don't use the `range` value capture, it will create an in-memory copy of the storage cluster - It is a very big structure

2. There is no need to create a new slice and start appending to it

```suggestion
	var found *ocsv1.StorageCluster = nil
	for i := range storageClusterList.Items {
		cluster := &storageClusterList.Items[i]
		if cluster.Spec.ExternalStorage.Enable {
			continue
		} 
		if found != nil {
			return nil, status.Errorf(codes.Internal, "found more than one storage cluster")
		}
		found = cluster
	}
	if found == nil {
		return nil, status.Errorf(codes.NotFound, "no storage cluster found")
	}
	return found
```

File: services/provider/server/server.go
Comment (nb-ohad) [inline services/provider/server/server.go:906]: Please add the new var to the end of the list
```suggestion
	var arr = []any{
		channelName,
		storageConsumer.Spec.StorageQuotaInGiB,
		inTransitEncryptionEnabled,
	}
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go]: Kept the getStorageCluster func as it's being used at another place too, created a separate isEncryptionInTransitEnabled() taking only the networkSpec.

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go]: Thanks for the detailed skeleton of the function, I have updated the function accordingly.

File: services/provider/server/server_test.go
Comment (rohan47) [inline services/provider/server/server_test.go:281]: Is this the test for encryption in transit is enabled?

Comment (malayparida2000) [reply services/provider/server/server_test.go:281]: No this just creates a storagecluster, as I am adding a util function in our code path which expects a storagecluster otherwise it will return an error. This is just to prevent that.

**Comment by nb-ohad** [comment]
@malayparida2000 Please let the reviewers resolve review comments threads. 
A reviewer must validate the changes made after you fix/change anything. We usually use the resolve state of a thread as confirmation that the change was made or the conversation resulted in a resolution that satisfies the reviewer, so the correct person to resolve a review comment thread is the reviewer after confirmation

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go]: Have addressed all the conversations along with this one, PTAL.

**Comment by malayparida2000** [comment]
/cc @leelavg 
Since Ohad will be on PTO until the 21st of October, could you please take up the responsibility for reviewing this PR? It's crucial that we have sufficient time after merging these changes to conduct end-to-end testing for the epic, identify any faults, and fix them. The current delay in the review process is causing us to lose valuable time. Additionally, I need to document the end-to-end workflow for all three modes, so your help with the review & merging the changes would be greatly appreciated.

File: services/provider/server/server.go
Comment (leelavg) [inline services/provider/server/server.go:898]: the implementation seems to be a roundabout way to me, in a namespace there'll only be 1 storagecluster isn't it? you could limit the list by 2 and anything that isn't 1 is an error scenario.

moreover from 4.18 api-server is always deployed and checking storagecluster for external enabled, the condition is implicitly missing out internal only scenario and I suggest storagecluster w/ allowremotestorageclusters atm is more suitable for the condition here.

**Comment by leelavg** [comment]
> help with the review & merging
- I can only review but don't have merge rights

> 21st of October
- can drop a note to review on 20th as well as it's working day for Ohad

**Comment by malayparida2000** [comment]
> I can only review but don't have merge rights
Right, But if you ack LGTM, I can get it merged.

> can drop a note to review on 20th as well as it's working day for Ohad
He told he is on PTO till 21st, so he should be back on 22nd only.

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go:199]: The error the getStorageCluster is returning is already formated in gRPC conventions so we can directly return it here instead of reformating it again.

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go:898]: Acc to how the current implimenetation is if there are multiple storageClusters, only one of them is handled, the others are just moved into Phase: Ignored. So I don't think erroring out ACC to the lits's size would be a good indicator.  

Agree, specifically looking for cluster with allowremotestorageclusters seems to be the best filter.

Comment (malayparida2000) [reply services/provider/server/server.go:899]: Updated the func

Comment (leelavg) [reply services/provider/server/server.go:899]: > into Phase: Ignored
- didn't know that, thanks!

> I don't think erroring out ACC to the lits's size would be a good choice.
- fair enough, then you should also be filtering out storagecluster that is set w/ ignored phase isn't it?

File: services/provider/server/server.go
Comment (leelavg) [reply services/provider/server/server.go:200]: ack, resolve comment.

File: services/provider/server/server.go
Comment (malayparida2000) [reply services/provider/server/server.go:899]: Right, Added a check for that.

**Review by iamniting — APPROVED** [review]
looks good to me
/hold only for @leelavg as @nb-ohad is not available

**Comment by malayparida2000** [comment]
/unhold
Resolving the hold as Leela is also on PTO, And this is required for testing in next week's build, Any follow on suggestions if there can be resolved with follow on PR.

**Pull request merged**
